File: playground/ios/Playground/Playground for AGenUI/Stability/StabilityLogger.swift
import Foundation

/// Writes stability test metrics in JSONL format.
/// Each round appends one JSON line and flushes to disk (updating mtime for heartbeat detection).
class StabilityLogger {
    private let outputDir: URL
    private let logFileURL: URL
    private var fileHandle: FileHandle?
    private let queue = DispatchQueue(label: "com.agenui.stability.logger")
    private let dateFormatter: DateFormatter = {
        let df = DateFormatter()
        df.dateFormat = "yyyy-MM-dd'T'HH:mm:ss.SSS"
        df.locale = Locale(identifier: "en_US_POSIX")
        return df
    }()

    var logFilePath: String { logFileURL.path }

    init(outputDir: URL) {
        self.outputDir = outputDir
        self.logFileURL = outputDir.appendingPathComponent("stability_log.jsonl")
        setupFile()
    }

    private func setupFile() {
        let fm = FileManager.default
        if !fm.fileExists(atPath: outputDir.path) {
            try? fm.createDirectory(at: outputDir, withIntermediateDirectories: true)
        }
        if !fm.fileExists(atPath: logFileURL.path) {
            fm.createFile(atPath: logFileURL.path, contents: nil)
        }
        // Append mode: launch.sh clears the file before first launch,
        // but crash-restarts by monitor.sh must preserve previous data.
        fileHandle = try? FileHandle(forWritingTo: logFileURL)
        fileHandle?.seekToEndOfFile()
    }

    func logRound(round: Int, scenario: String, durationMs: Int, status: String,
                  fixture: String?, error: String?, memoryMb: Double) {
        var dict: [String: Any] = [
            "ts": dateFormatter.string(from: Date()),
            "round": round,
            "scenario": scenario,
            "duration_ms": durationMs,
            "memory_mb": Double(Int(memoryMb * 10)) / 10.0,
            "status": status
        ]
        if let f = fixture { dict["fixture"] = f }
        if let e = error { dict["error"] = e }

        writeLine(dict)
    }

    func logEvent(event: String, detail: String) {
        let dict: [String: Any] = [
            "ts": dateFormatter.string(from: Date()),
            "event": event,
            "detail": detail
        ]
        writeLine(dict)
    }

    private func writeLine(_ dict: [String: Any]) {
        guard let data = try? JSONSerialization.data(withJSONObject: dict, options: [.sortedKeys]),
              var line = String(data: data, encoding: .utf8) else { return }
        line += "\n"
        guard let lineData = line.data(using: .utf8) else { return }

        queue.sync {
            self.fileHandle?.write(lineData)
            self.fileHandle?.synchronizeFile()
        }
    }

    func close() {
        queue.sync {
            self.fileHandle?.synchronizeFile()
            self.fileHandle?.closeFile()
            self.fileHandle = nil
        }
    }

    deinit {
        close()
    }
}

File: playground/android/app/src/main/java/com/amap/agenuiplayground/stability/StabilityLogger.java
package com.amap.agenuiplayground.stability;

import android.os.Debug;

import java.io.File;
import java.io.FileWriter;
import java.io.IOException;
import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.Locale;

/**
 * JSONL logger for stability test metrics.
 * Writes one JSON line per round to a local file for later analysis.
 */
public class StabilityLogger {
    private final File logFile;
    private FileWriter writer;
    private final SimpleDateFormat dateFormat =
            new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSS", Locale.US);

    public StabilityLogger(File outputDir) {
        this.logFile = new File(outputDir, "stability_log.jsonl");
        try {
            outputDir.mkdirs();
            // Append mode: launch.sh clears the file before first launch,
            // but crash-restarts by monitor.sh must preserve previous data.
            writer = new FileWriter(logFile, true);
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    public void logRound(int round, String scenario, long durationMs, String status,
                         String fixture, String error) {
        long nativeHeap = Debug.getNativeHeapAllocatedSize() / (1024 * 1024);
        long javaUsed = (Runtime.getRuntime().totalMemory() - Runtime.getRuntime().freeMemory()) / (1024 * 1024);
        double memoryMb = nativeHeap + javaUsed;

        StringBuilder sb = new StringBuilder();
        sb.append("{\"ts\":\"").append(dateFormat.format(new Date())).append("\"");
        sb.append(",\"round\":").append(round);
        sb.append(",\"scenario\":\"").append(escapeJson(scenario)).append("\"");
        sb.append(",\"duration_ms\":").append(durationMs);
        sb.append(",\"memory_native_mb\":").append(nativeHeap);
        sb.append(",\"memory_java_mb\":").append(javaUsed);
        sb.append(",\"memory_total_mb\":").append(String.format(Locale.US, "%.1f", memoryMb));
        sb.append(",\"status\":\"").append(status).append("\"");
        if (fixture != null) {
            sb.append(",\"fixture\":\"").append(escapeJson(fixture)).append("\"");
        }
        if (error != null) {
            sb.append(",\"error\":\"").append(escapeJson(error)).append("\"");
        }
        sb.append("}\n");

        writeLine(sb.toString());
    }

    public void logEvent(String event, String detail) {
        StringBuilder sb = new StringBuilder();
        sb.append("{\"ts\":\"").append(dateFormat.format(new Date())).append("\"");
        sb.append(",\"event\":\"").append(escapeJson(event)).append("\"");
        if (detail != null) {
            sb.append(",\"detail\":\"").append(escapeJson(detail)).append("\"");
        }
        sb.append("}\n");
        writeLine(sb.toString());
    }

    private synchronized void writeLine(String line) {
        if (writer != null) {
            try {
                writer.write(line);
                writer.flush();
            } catch (IOException e) {
                e.printStackTrace();
            }
        }
    }

    public void close() {
        if (writer != null) {
            try {
                writer.close();
            } catch (IOException e) {
                e.printStackTrace();
            }
        }
    }

    public String getLogFilePath() {
        return logFile.getAbsolutePath();
    }

    private static String escapeJson(String s) {
        if (s == null) return "";
        return s.replace("\\", "\\\\")
                .replace("\"", "\\\"")
                .replace("\n", "\\n")
                .replace("\r", "\\r")
                .replace("\t", "\\t");
    }
}
